### ISSUE TITLE
Port token replacements to Rust, improve token handling

### ISSUE THREAD

**Issue by lizziegooding** [ingalls/pt2itp#457]
## Phase 1

Write a replace function for the fancy-regex lib that will allow us to find groups within strings and replace them with new strings. This should be a new module within the /text processing folder of the Rust code in pt2itp. 

## Phase 2

The pt2itp JavaScript code should never import geocoder-abbreviations. Right now, we import geocoder-abbreviations in the map module and then pass it to the Rust code. All Rust code that needs the tokens should instead import that code itself. For example, the `context` struct should call geocoder-abbreviations and store all tokens itself (in the `new` call).

## Phase 3

Remove support for the `tokenless` property on features. Rather than a string, the `tokenized` property should be a vector of structs like `"tokenized": [ { "token": "main", "type": "unknown }, { "token": "st", "type": "type" }]`. this will allow us to generate `tokenless` on the fly by dropping all tokens which have a `unknown` or non-street type.

**Comment by lizziegooding** [comment]
## [phase 4]

Add a final phase (this was originally implied as part of phase 3 but probably makes sense to break out separately, we'll add support for regex token replacements using the fancy-regex crate geocoder-abbreviations uses. This will require wrapping some fancy-regex types into custom types and implementing `PartialEq` and potentially other traits.

**Comment by lizziegooding** [comment]
Phases 1-3 are written and awaiting additional manual review https://github.com/ingalls/pt2itp/pull/466. For this first version

*conflate will not add features where*:
- one feature is missing a road type
- one feature is missing a cardinal
- the cardinals match but are in a different position
- the road types match but are in a different position

*conflate will still add features where*:
- both features contain road types and they don't match
- both features contain cardinals and they don't match

We may ship phases 1-3 first and implement phase 4 separately as soon as we're confident in phase 3.

**Comment by lizziegooding** [comment]
Posting a more complete summary of @ingalls and my work to date on conflate and token handling.

*tl;dr* the new conflate mode is correctly detecting duplicates and has dramatically reduced the number of duplicate features we're adding to Hecate. There are still known cases in which we are erroneously adding duplicate features.

## How conflate works

Conflate mode allows us to compare a new dataset with the data we already have in Hecate to ensure that we're not adding duplicate features. It performs two main actions; `create`, in which we add a new feature to Hecate and `modify`, in which we update an existing feature. Right now, the only property we're updating is alternative names (stored in the `properties.street` array).

Unlike on master, the new conflate mode will detect an existing address in Hecate in the following cases:
- one feature is missing a road type (e.g. Loveland: `1416 Santa Cruz`, Hecate: `1416 SANTA CRUZ DR`)
- one feature is missing a cardinal (e.g. Loveland: `3038 Country Club Rd E`, Hecate: `3038 COUNTRY CLUB RD`)
- cardinals match but are in a different position (e.g. Loveland: `15275 W FM 968`, Hecate `15275 FM 968 W`)
- road types match but are in a different position (e.g `Main St` vs `St Main`)
Conflate will `modify` the existing address rather than `create` a duplicate address.

The new conflate mode will still add duplicate addresses in the following cases:
- features have mismatching road types (e.g. Loveland: `2500 ALPINE RD`, Hecate: `2500 Alpine St`).
- features have mismatching cardinals (e.g. `Main St E` vs `Main St W`)
- features have differing whitespace (e.g. Loveland: `100 pine crest dr`, Hecate: `100 Pinecrest Dr`)

We are intentionally preventing conflate from detecting duplicates where cardinals and road types don't match because we want to err on the side of stricter matching and adding more points where possible. While this may add some additional noise, there are many instances where distinct proximate streets do have different cardinals or road types and should be added as new features.

![Screen_Shot_2019-05-22_at_12_12_42_PM](https://user-images.githubusercontent.com/18095166/58570721-77a18700-8206-11e9-9bed-6748d43424e5.png)
_72nd Dr, Av, and Rd in Queens_

We should handle whitespace differences in the future. We have not yet scoped this task and it's not on our immediate roadmap.

## Stats summary

To compare the effectiveness of the new conflate mode vs master, I reran conflate using a Loveland source from Harrison County Texas. The original Loveland source contained 890 parcels.

`master`
Features added: 91
Features modified: 0

`conflate`
Features added: 25
Features modified: 32

During @samely's initial review, she found that there were only 11 new features that should actually be added of the 91 detected by conflate on `master`. We don't have a list of those exact features, but after reviewing the output of the new conflate, I found that 15 of the 25 addresses conflate identified were valid. Of the invalid features, 7 were not detected as duplicates due to known whitespace issues, and 3 appear to be invalid or non existant addresses, indicating poor quality Loveland data.

Anecdotally, the Loveland addresses we would add in this county don't appear to be particularly high value. Many of them are empty lots and only a handful appear to be commercial properties. Harrison county is quite rural and small (only 66,000 people), so this may not be the case in other counties.

**Comment by lizziegooding** [comment]
Keeping this open as we still haven't shipped support for regex token replacements https://github.com/mapbox/pt2itp/pull/476
